Mirror global properties in a child libctx

mattcaswell opened the pull request:
If using a child library context that the global properties from the application library context should be mirrored in the child.

t8m commented:
Looking at the CI runs it is interesting how none of the new runchecker builds found anything where the original CI runs fail quite wildly. Nice indication that our original CI runs are well tuned to find issues.

t8m commented:
Although this time it is just because all the new runchecker jobs just do not have enable-fips. :grin:

mattcaswell commented:
Updated with a fix for "enable-fips" and updating the checksums. Lets see if the CIs are any happier now.

mattcaswell commented:
Hmmmm. @t8m, I rebased this PR and removed all updates to the fips checksums to avoid conflicts in the checksums files. I was expecting that I'd get the new label applied to this PR...but it doesn't seem to have been added?

t8m commented:
Yep, the dependency does not work properly and the checksums get regenerated. Fix in #15267

t8m commented:
Hooray! Works. If you click on details of the job and look at the make diff-fips-checksums step log it shows the affected files nicely.

mattcaswell commented:
Rebased again. Labelling is now correct.

Please take a look.

mattcaswell commented:
Ping?

mattcaswell commented:
Rebased to resolve conflicts with master.

Ping, ping!!

paulidale commented on crypto/property/property_parse.c:
```diff
@@ -616,3 +616,133 @@ int ossl_property_parse_init(OSSL_LIB_CTX *ctx)
+        **buf = '\0';
+    else
+        **buf = ch;
+    *buf += 1;
```
What about `++*buf;` here and below?

paulidale commented on crypto/property/property_parse.c:
```diff
@@ -616,3 +616,133 @@ int ossl_property_parse_init(OSSL_LIB_CTX *ctx)
+    if (*remain == 0)
+        return;
+
+    BIO_snprintf(*buf, *remain, "%d", val);
```
`BIO_snprintf` range checks, is the early checking code required?

mattcaswell replied:
We  still need to calculate the length that *would* be required if the buffer was big enough. So I don't think the BIO_snprintf range checks help.

paulidale commented on doc/man3/OSSL_LIB_CTX.pod:
```diff
@@ -69,6 +69,12 @@ reference count. L<OSSL_PROVIDER_unload(3)> must not be called for a provider in
 
+In addition to providers, a child library context will also mirror the default
+properties (set via L<EVP_set_default_properties(3)>) from the parent library
+context. If L<EVP_set_default_properties(3)> is called directly from a child
```
_from_ or _on_?

paulidale commented on doc/man7/provider-base.pod:
```diff
@@ -289,12 +290,16 @@ I<create_cb> is a callback that will be called when a new provider is loaded
+return 1 on success or 0 on failure.
+
+I<global_props_cb> is a callback that will be called when the global properties
+from the parent library context are changed.It should return 1 on success
```
Space after full stop.

paulidale reviewed (approved):
A few minor items.

mattcaswell commented:
Updated pushed addressing feedback. Please reconfirm

t8m commented:
The CI failures might be relevant as I do not see them in other runs

mattcaswell commented:
> The CI failures might be relevant as I do not see them in other runs

Hmm. When it fails it seems to be in the endecode_test. But I cannot reproduce this locally at all. I suspect it may be an intermittent failure - which is why it seems to be affecting various builds fairly randomly. Not sure how to progress this.

mattcaswell commented:
I found a potential issue which explains why a lot of the CI tests failed to complete. Not sure if it explains the crashes or not. Lets see how the CIs do now.

t8m commented:
Unfortunately CI is offline now. :disappointed:

paulidale commented on crypto/property/property_parse.c:
```diff
@@ -616,3 +616,133 @@ int ossl_property_parse_init(OSSL_LIB_CTX *ctx)
+static void put_char(char ch, char **buf, size_t *remain, size_t *needed)
+{
+    if (*remain == 0) {
+        *needed += 1;
```
Missed one :)

paulidale commented:
No need to reset the timer.

mattcaswell commented:
Pushed. Thanks.
